design/decisions/0008-extensible-config-management.md

```diff
@@ -0,0 +1,193 @@
+# Extensible configuration management
+
+* Date: __2025-08-12__
+* Status: __New__
+
+## Background
+
+The built-in configuration and log management features offered by `tedge-agent` has the following limitations:
+
+- only file based configurations and logs can be retrieved.
+- while updating a configuration file, there are no pre/post actions that can be performed.
+  For e.g, a user would want to backup the existing configuration before updating the configuration,
+  and restart a service after the configuration is updated.
+- Though the tedge workflows feature allows overriding the existing behavior,
+  when a built-in operation is overridden, you end up losing all existing functionality,
+  like uploading/downloading the files from the cloud as well.
+
+The goal is to provide an interface that allows users to extend only the specific aspect of the operation that they want,
+like the pre-update or post-update actions, leaving the rest of the behavior unchanged.
+
+## Solution proposal
+
+- The agent supports plugins for log and config management.
+- The supported types for the respective operation is gathered from all these plugins.
+- The agent maintains a mapping of the types and their corresponding plugins.
+- When an an operation is received for a type, it is delegated to the corresponding plugin.
+
+### Log Management
+
+- `tedge-agent` to add support for multiple log management plugins defined under `/etc/tedge/log-plugins`.
+- Migrate the existing file-based log functionality into a dedicated `file` plugin which is available by default.
+- To retrieve logs from other sources like systemd journal or docker, add respective plugins for these sources.
+- Each plugin must support the following sub-commands:
+  - `list`: list all the log types that it supports.
+    Used to detect if this is a valid plugin if it exits with exit code 0.
+    The output is used only when auto-discovery for log types is turned ON for this plugin.
+  - `get <type> <temp-log-file-path> [--from <timestamp>] [--to <timestamp>] [--filter <filter-text>]`:
+    Used to fetch the log for the given type within the provided log range.
+    The log content must be written to the temporary file passed provided by the agent.
+- The plugin directory would also have a `conf.d` directory where each plugin can store their configuration in a toml file.
```

Comment:
Sorry, I'm lost here. Why do we need this config settings? What does mean `auto_discover = false`? Isn't it enough to add `.ignore` extension or whatever to the plugin like `tedge diag collect` does? Why useful to have `include` and `exclude` settings?

Reply:
Thanks for explaining. I understood from your comment and the other thread. I prefer `use_list` as it sounds clearer.

Reply:
Yes. When `auto_discover` is off, when a new log entry has to be added, its `include` entry must be added to the config, most-likely via a drop-in extension config at `/etc/tedge/log-plugins.d`.